### PR TITLE
Adopt MLUtils.DataLoader for minibatch iteration

### PR DESCRIPTION
**Pull request by cossio**
Explores what adopting MLUtils.jl for minibatch iteration looks like, replacing the hand-rolled machinery in `src/train/infinite_minibatches.jl` (net −34 lines; the file drops from 77 to 42 lines).

## What changed

- `infinite_minibatches` now wraps `Iterators.cycle(DataLoader(...; batchsize, shuffle, partial = false))`. Cycling restarts the loader when an epoch is exhausted, and the loader reshuffles on each restart, reproducing the old infinite stream with a fresh permutation per epoch; `partial = false` reproduces the old fixed-size batches (trailing partial batch dropped).
- The internal helpers `nobs`, `getobs`, `shuffleobs`, and `InfiniteMinibatchIterator` are removed (none were marked `public`). `_prepare_training_data` selects the positive-weight subset with MLUtils `getobs` on a `(data, wts)` tuple.
- MLUtils becomes a direct dependency (compat `0.4`).
- `randperm` is no longer imported (its only caller was `shuffleobs`).

## The `nothing`-weights seam

MLUtils deliberately defines no observation semantics for `Nothing` (and extending `numobs`/`getobs` to it from here would be type piracy), so the old iterator's `nothing`-threading is handled by dispatch at the call boundary: the `wts::AbstractVector` method batches the `(data, wts)` tuple jointly, and the `wts::Nothing` method batches `data` alone and pairs each batch with `nothing`, so the training loop keeps destructuring `(vd, wd)` unchanged.

## One semantic difference, guarded

`DataLoader` **clamps** `batchsize > numobs` down to the dataset size (with a warning), whereas the old iterator yielded an empty stream (i.e. zero training iterations). To keep this PR a pure refactor, the wrapper guards that case and returns an empty iterator, preserving the historical contract. If the clamping behavior is actually preferred (arguably the silent no-op is a footgun — note `_prepare_training_data` already clamps `batchsize` to the positive-weight count on the weighted path, so the two paths disagree today), that guard can be dropped in a follow-up.

## Trade-off to weigh

This trades ~60 lines of dependency-free iterator code for MLUtils' dependency chain (NNlib, StatsBase, Tables, MLCore, …), which is heavier than anything currently in `[deps]`. The alternative discussed was extracting the old iterator into a micro-package instead. This PR exists to make that comparison concrete.

Internal mechanics also improve slightly: the old iterator materialized a full shuffled copy of the dataset every epoch, while `DataLoader` keeps an index permutation and gathers per batch.

## Tests

- `test/infinite_minibatches.jl` rewritten against the public behavior (cycling, joint data/weight permutation, per-epoch reshuffle, batchsize edge cases) instead of the removed internals.
- Verified locally on Julia 1.12.6: infinite_minibatches, train, pcd, zero_weight_training, jlarrays (JLArrays with `allowscalar(false)` — the DataLoader gather path stays GPU-safe), ExplicitImports, and Aqua all pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01LtVghvmoChrEuTxuZ75Nsd

---
_Generated by [Claude Code](https://claude.ai/code/session_01LtVghvmoChrEuTxuZ75Nsd)_

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,16 @@
 
 All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
```
[the remark this answers is not in the document]
Comment (cossio) [reply CHANGELOG.md:3]: Done in 4c8e92b: `Project.toml` now reads `version = "6.3.0-DEV"`. Chose the minor-bump dev version by the precedent of v6.1.0, where the analogous LogStatFunctions adoption (new dependency plus removal of non-`public` helpers) was released as a minor version; the final release number remains the owner's call at release time.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/infinite_minibatches.jl [hunk 2]
```diff
@@ -1,59 +1,49 @@
...
-    data = randn(2, 5)
-    @test iterate(infinite_minibatches(data; batchsize = 6, shuffle = false)) === nothing
-    @test iterate(infinite_minibatches(nothing; batchsize = 1, shuffle = false)) === nothing
+@testset "shuffle draws a fresh permutation each epoch" begin
+    orders = Set{Vector{Int}}()
+    for (x, w) in Iterators.take(infinite_minibatches(1:10, collect(1:10); batchsize = 10, shuffle = true), 20)
+        @test sort(x) == 1:10
+        @test x == w
+        push!(orders, copy(x))
+    end
```
[the remark this answers is not in the document]
Comment (cossio) [reply test/infinite_minibatches.jl:38]: This isn't how Julia's hashing containers behave: `Base.hash(::AbstractArray)` and `isequal` are defined element-wise, so a `Set{Vector{Int}}` deduplicates by content, not object identity. Verified concretely: `push!`ing `[1,2,3]`, `copy([1,2,3])`, and `[3,2,1]` into a `Set{Vector{Int}}` yields `length == 2`. Repeated permutations therefore collapse to one entry and the reshuffle assertion is effective as written. No change needed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/train/infinite_minibatches.jl [hunk 3]
```diff
@@ -1,64 +1,29 @@
...
-        return iterate(iter) # restart iteration
-    else
-        items = getobs(i:(i + iter.batchsize - 1), shuffled...)
-        return items, (i + iter.batchsize, shuffled)
-    end
+function infinite_minibatches(
+        data::AbstractArray, wts::AbstractVector; batchsize::Int, shuffle::Bool = true
+    )
+    batchsize > 0 || throw(ArgumentError("batchsize must be positive"))
+    batchsize ≤ numobs((data, wts)) || return ()
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/train/infinite_minibatches.jl:13]: Intentional: `infinite_minibatches` is not marked `public`, and since v6.0.0 this package explicitly guarantees compatibility only for `public` names — internal helpers may change without a breaking release (see the v6.0.0 CHANGELOG entry). The only caller in the package always passed both arguments, and the signature change is recorded in the CHANGELOG alongside the removed helpers. Keeping variadic forwarding methods would preserve exactly the `nothing`-threading surface this PR is meant to retire, so I'd rather not add them.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (cossio) [reply CHANGELOG.md:3]: Update: master has since opened the dev cycle as `6.2.1-DEV` (#210), which conflicted with the `6.3.0-DEV` set here. Resolved to master's `6.2.1-DEV` in the merge commit — the tree no longer identifies as the released 6.2.0, which was the substance of this finding, and the final release number (likely minor, given the new dependency) stays with the release run.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,14 @@ All notable changes to this project will be documented in this file. The format
 
 ## Unreleased
 
+- Minibatch iteration during training is now backed by
+  [MLUtils.jl](https://github.com/JuliaML/MLUtils.jl), which becomes a new
+  dependency: `infinite_minibatches` cycles an `MLUtils.DataLoader`, replacing
```
[the remark this answers is not in the document]
Comment (cossio) [reply CHANGELOG.md:9]: Declining this one: keeping master's `6.2.1-DEV` matches the repository's actual practice. In the previous cycle the version stayed at `6.1.1-DEV` the entire time while minor-worthy changes accumulated (the moments-array API rework, `grad2ave`/`grad2var` removal), and the release commit went directly `6.1.1-DEV → 6.2.0` — a `6.2.0-DEV` never existed in history. The `-DEV` number is the patch-next placeholder opened by the release run, not a live tracker of the eventual release level; the release skill chooses the real number from the CHANGELOG, and CLAUDE.md leaves that decision to the owner. Overriding the owner's fresh dev-cycle bump (#210) inside this feature PR would invert that flow. The earlier `4c8e92b` was superseded by exactly that consideration when the merge conflict surfaced it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by cossio** [comment]
Superseded by #222, which rebuilds this change on top of #220 (merging first). Stacking on #220 dissolves both complications here: the `nothing`-weights dispatch seam (weights are always arrays there) and the empty-stream guard (#220's trainers clamp oversized batchsizes, which is DataLoader's native behavior). All review threads here were resolved; the discussion carries over.

---
_Generated by [Claude Code](https://claude.ai/code)_
